src/main/java/software/amazon/awssdk/eventstreamrpc/OperationContinuationHandler.java

```diff
@@ -189,128 +199,155 @@ final public CompletableFuture<Void> sendStreamEvent(final StreamingResponseType
+                    e.getClass().getCanonicalName(), e.getMessage());
         }
     }
 
     @Override
-    final protected void onContinuationMessage(List<Header> list, byte[] bytes, MessageType messageType, int messageFlags) {
+    final protected void onContinuationMessage(List<Header> list, byte[] bytes, MessageType messageType,
+                                               int messageFlags) {
         LOGGER.debug("Continuation native id: " + continuation.getNativeHandle());
 
         //We can prevent a client from sending a request, and hanging up before receiving a response
         //but doing so will prevent any work from being done
         if (initialRequest == null && (messageFlags & MessageFlags.TerminateStream.getByteValue()) != 0) {
-            LOGGER.debug("Not invoking " + getOperationName() + " operation for client request received with a " +
-                    "terminate flag set to 1");
+            LOGGER.debug("Not invoking " + getOperationName() + " operation for client request received with a "
+                    + "terminate flag set to 1");
             return;
         }
         final EventStreamRPCServiceModel serviceModel = getOperationModelContext().getServiceModel();
         try {
             if (initialRequest != null) {
                 // Empty close stream messages from the client are valid. Do not need any processing here.
-                if ((messageFlags & MessageFlags.TerminateStream.getByteValue()) != 0 && (bytes == null || bytes.length == 0)) {
+                if ((messageFlags & MessageFlags.TerminateStream.getByteValue()) != 0 && (bytes == null
+                        || bytes.length == 0)) {
                     return;
                 } else {
                     final StreamingRequestType streamEvent = serviceModel.fromJson(getStreamingRequestClass(), bytes);
                     //exceptions occurring during this processing will result in closure of stream
                     handleStreamEvent(streamEvent);
                 }
-            } else { //this is the initial request
+            } else {
+                //this is the initial request
                 initialRequestHeaders = new ArrayList<>(list);
                 initialRequest = serviceModel.fromJson(getRequestClass(), bytes);
                 //call into business logic
-                final ResponseType result = handleRequest(initialRequest);
-                if (result != null) {
-                    if (!getResponseClass().isInstance(result)) {
-                        throw new RuntimeException("Handler for operation [" + getOperationName()
-                                + "] did not return expected type. Found: " + result.getClass().getName());
+                CompletableFuture<ResponseType> resultFuture = handleRequestAsync(initialRequest);
+                if (resultFuture == null) {
+                    resultFuture = CompletableFuture.completedFuture(handleRequest(initialRequest));
```

Comment:
the old implementation would `throw new RuntimeException("Operation handler returned null response!");` if this returned null, the new implementation will just `NullPointerException` in the next few lines ... which is fine. But just pointing it out

Reply:
No, this is different. The null completable future is part of the specification of the method. If the completable future _resolves_ to null, then the same exception will be logged that the handler returned null.